pkg/app/api/grpcapi/piped_api.go

```diff
@@ -1206,13 +1210,33 @@ func (a *PipedAPI) InChainDeploymentPlannable(ctx context.Context, req *pipedser
 		return nil, status.Error(codes.InvalidArgument, "invalid deployment with chain block index provided")
 	}
 
-	prevBlockSuccessfullyCompleted, err := dc.IsSuccessfullyCompletedBlock(req.DeploymentChainBlockIndex - 1)
-	if err != nil {
-		return nil, status.Error(codes.InvalidArgument, "unable to process the previous block of this deployment in chain")
+	previousBlock := dc.Blocks[req.DeploymentChainBlockIndex-1]
+	// If the previous block has not finished yet, should not plan this deployment to run.
+	if !previousBlock.IsCompleted() {
+		return &pipedservice.InChainDeploymentPlannableResponse{
+			Plannable: false,
+		}, nil
+	}
+
+	var (
+		plannable, cancel bool
+		reason            string
+	)
+	switch previousBlock.Status {
+	case model.ChainBlockStatus_DEPLOYMENT_BLOCK_SUCCESS:
+		plannable = true
+	case model.ChainBlockStatus_DEPLOYMENT_BLOCK_FAILURE:
+		cancel = true
+		reason = "Previous block finished with FAILURE status"
+	case model.ChainBlockStatus_DEPLOYMENT_BLOCK_CANCELLED:
+		cancel = true
+		reason = "Previous block finished with CANCELLED status"
```

Comment:
I wonder if it should show reason why it's not plannable even if it's other statuses like `UNKNOWN` and `RUNNING`. Let's say the previous block is stuck with the `UNKNOWN` status, you could be confused as everything stops without any log.

I understand the response structure could get complicated to do that, but what about you?

```suggestion
		reason = "Previous block finished with CANCELLED status"
	default:
	        reason = fmt.Sprintf("not plannable because the status of previous block is %v", previousBlock.Status)
```

Reply:
That reason string contains the value for CancelReason, and in case the status other than FAILURE | CANCELLED we do not cancel the deployment. So I don't think we need to do that. Besides, the status UNKNOWN is just a fail case defined. Basically, in the current codebase, there is no way it could be blocked status = UNKNOWN, I thought about removing that UNKNOWN status, but we may need it as value to return in some failed case in the future, so I kept it. How do you think about remove that status to reduce mislead 🤔

Reply:
I meant it's confusing to show error message without any reason when it's not plannable: https://github.com/pipe-cd/pipe/pull/2866/files#diff-38efeefa59e26254919993c1d15e9177559a52f3b8786035821c7e1e5e7131c8R396

But look closely, the severity is Info, so never mind!

>How do you think about remove that status to reduce mislead 🤔

I'm gonna say it's fine to be kept as a default value.

Reply:
Gotcha 🙆‍♀️

Reply:
Yes, `UNKNOWN` could be considered as a fail case (nice point), but I prefer to sync with the deployment status so removing it is fine since the deployment status does not contain that kind.

Reply:
Removed this UNKNOWN status too 👍